docs/prover/approx/grounding.md

```diff
@@ -0,0 +1,267 @@
+
+```
+rule recursiveQuantifier {
+    require forall mathint x . f(x) > f(x-1);
+    assert f(8) > f(6);
+}
+```
+
+Although we can see that the assertion must be true, we would need to combine
+the statements `f(8) > f(7)` and `f(7) > f(6)` to prove it, and the grounding
+mechanism is unable to do this.
+
+```{todo}
+What does the counterexample look like?
+```
+
+```{warning}
+The Prover will not detect recursive quantified statements like this, so this
+would be something good to check for when you encounter spurious counterexamples.
+```
+
+### Variables That Are Not Arguments
+
+In order for grounding to work, every variable appearing in a quantified
+statement must be used as an argument to a function.  For example, neither of
+the following examples will work:
+
+```{todo}
+"must be used" or "must only be used"?
+```
+
+```
+require forall mathint x . x * 2 != y;
+require forall mathint x . forall mathint y . f(x) != 2y;
+```
+
+Grounding is based on the application of functions, so the grounded variables
+must be arguments to functions.  In this case, `x` and `y` are not arguments to
+functions.
+
+In such cases, we will hard crash, so again you can rewrite your spec or turn
+grounding off. To avoid this, ensure that each quantified variable is used as
+an argument of a function call within the quantified expression.
+
+```{todo}
+Crash how?
+```
+
+```{todo}
+What is the following example showing?
+```
+
+This may not
+always be possible, but we can write:
+
+```
+forall x. f(x) % 2 == 1
+```
+ 
+### Double Polarity
+
+The polarity of a sub-formula is the direction of the effect it has on the
+output of the formula. This is best demonstrated through an example:
+
+```cvl
+a && (b || !c)
+```
+
+In this example, `b` possesses a positive polarity because change `b` from `false` to `true`
+can't make the formula `false` if it wasn't before.  Informally, making `b`
+"more true" can only make the formula "more true".
+
+In this example, `a` also possesses positive polarity: if the formula was true when `a` was
+`false`, it must also be `true` when `a` is `true`.
+
+On the other hand, `c` has a negative polarity because changing `c` from `true`
+to `false` can only make the statement "more false".  If can't become `true` if
+it wasn't true before.
+
+Sub-expressions can also have double polarity.  For example, consider the
+formula
+```
+a <=> b
+```
+In this example, `a` has double polarity, because making `a` true could cause
+the formula to become `false` when it was true before, but it could also cause
+the formula to become `true` when it wasn't before.
+
+Grounding crashes when the variable in the quantifier has double polarity in
+the quantified expression.  In that case, you can either rewrite the quantified
+statement or {ref}`turn grounding off <-smt_groundQuantifiers>`.
+
+```{todo}
+Crashes how?  Error message?
+```
+
+```{todo}
+I don't understand the following examples
+```
+
+```
+ite(forall .., x, y)
+```
+
```

Comment:
here the `forall` statement is used as the condition of an ite - that means it acts with both positive and negative polarity in the eventual formula. Easiest way to see this is looking at an `ite` in cnf form (let's say its all just boolean):
```
ite(i, t, e) = (i /\ t) \/ (!i /\ e)
```
So `i` appears both positively and negatively in the formula.